srml/system/src/offchain.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Module helpers for offchain calls.
+
+use codec::Encode;
+use sr_primitives::app_crypto::RuntimeAppPublic;
+use sr_primitives::traits::Extrinsic as ExtrinsicT;
+
+/// A trait responsible for signing a payload using given account.
+pub trait Signer<Account, Signature> {
+	/// Sign any encodable payload with given account and produce a signature.
+	///
+	/// Returns `Some` if signing succeeded and `None` in case the `account` couldn't be used.
+	fn sign<Payload: Encode>(account: Account, payload: &Payload) -> Option<Signature>;
+}
+
+impl<Account, Signature, AppPublic> Signer<Account, Signature> for AppPublic where
+	AppPublic: RuntimeAppPublic + From<Account>,
+	AppPublic::Signature: Into<Signature>,
```

Comment:
I am a bit unclear on these new crypto types: what do the bounds here translate to? Who will end up having this impl?

Reply:
My understanding:

So we have a low-level crypto types (like `sr25519` and `ed25519`) that are commonly used within the "core" runtime part (as `AccountId`, etc).

We also have application-specific crypto types, which are basically wrappers for the low-level types with additional information about the module (i.e. application) they are coming from.

So for instance, `im-online` is (or may be) using `sr25519` as the crypto type, but instead of `sr25119::Public` you will rather use `im_online::app::Public`, as this type has additional `APP_ID` identification to know it's coming from `im_online` module.
The `APP_ID` is very useful especially when querying the `Keystore` cause we might have many sr25519 keys, but only one (or some) of them will be the one used by `im_online` module.


So the bounds here basically say:
> Please let me know the application-specific crypto type that you want to use for a particular `Account`.

Now that I think of it, in the future it might actually be better to use `AppPublic` instead of `Account` right from the start (i.e. when we initiate the signing), rather then convert in the middle of the process.
I'll refactor this if I find it more convenient to use that way, when equivocations reporting will be implemented.